azure/ConsiderationsForServiceDesign.md

```diff
@@ -0,0 +1,97 @@
+ - Building apps that consume the API
+ - Hold reviews and share what is learned with your team
+ - Get customer feedback from API previews
+ - Thinking about the code that a customer writes both before and after an HTTP operation
+ - Initializing and reading from the data structures your service requires
+ - Thinking about which errors are recoverable at runtime as opposed to indicating a bug in the customer code that must be fixed
+
+The whole purpose of a preview to address feedback by improving abstractions, naming, relationships, API operations, and so on. It is OK to make breaking changes during a preview to improve the experience now so that it is sustainable long term. 
+
+> :white_check_mark: **DO** provide examples in multiple languages
+> 
+> :white_check_mark: **DO** include at least one dynamically typed language (for example, Python or JavaScript) and one statically typed language (for example, Java or C#).
+
+### Focus on Hero Scenarios
+It is important to realize that writing an API is, in many cases, the easiest part of providing a delightful developer experience. There are a large number of downstream activities for each API, e.g. testing, documentation, client libraries, examples, blog posts, videos, and supporting customers in perpetuity. In fact, implementing an API is of miniscule cost compared to all the other downstream activities. 
+
+*For this reason, it is <b>much better</b> to ship with fewer features and only add new features over time as required by customers.*
+
+Focusing on hero scenarios reduces development, support, and maintenance costs; enables teams to align and reach consensus faster; and accelerates the time to delivery. A telltale sign of a service that has not focused on hero scenarios is "API drift," where endpoints are inconsistent, incomplete, or juxtaposed to one another. 
+
+> :white_check_mark: **DO** define "hero scenarios" first including abstractions, naming, relationships, and then define the API describing the operations required
+> 
+> :white_check_mark: **DO** provide example code demonstrating the "Hero Scenarios"
+> 
+> :no_entry: **DO NOT** proactively add APIs for speculative features customers might want 
+
+### Start with your API Definition
+Understanding how your service is used and defining its model and interaction patterns--its API--should be one of the earliest activities a service team undertakes. It reflects the abstractions & naming decisions and makes it easy for developers to implement the hero scenarios.  
+
+> :white_check_mark: **DO** provide an [OpenAPI Definition](https://swagger.io/specification/) (with [autorest extensions](https://github.com/Azure/autorest/blob/master/docs/extensions/readme.md)) describing the service. The OpenAPI definition is a key element of the Azure SDK plan and is essential for documentation, usability and discoverability of services.
```

Comment:
We should be cautious to clarify Swagger or OpenAPIv2, since the link points to v3. We do not support v3 as of today.

Reply:
Should we point to https://github.com/Azure/azure-rest-api-specs in the spirit of "if you want to lean more about Swagger authoring,  go here" ?

Reply:
Agreed.  We should point to the OpenAPI v2 spec.

Reply:
+1